cryptid_anchor/programs/middleware/check_recipient/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+extern crate core;
+
+use anchor_lang::prelude::*;
+use cryptid_anchor::cpi::accounts::ApproveExecution;
+use cryptid_anchor::program::CryptidAnchor;
+use cryptid_anchor::state::transaction_account::TransactionAccount;
+
+declare_id!("midcHDoZsxvMmNtUr8howe8MWFrJeHHPbAyJF1nHvyf");
+
+#[program]
+pub mod check_recipient {
+    use super::*;
+
+    pub fn create(ctx: Context<Create>, recipient: Pubkey, nonce: u8, bump: u8) -> Result<()> {
+        ctx.accounts.middleware_account.recipient = recipient;
+        ctx.accounts.middleware_account.authority = *ctx.accounts.authority.key;
+        ctx.accounts.middleware_account.nonce = nonce;
+        ctx.accounts.middleware_account.bump = bump;
+        Ok(())
+    }
+
+    pub fn execute_middleware(ctx: Context<ExecuteMiddleware>) -> Result<()> {
+        let transaction_account = &ctx.accounts.transaction_account;
+
+        // there can only be one instruction
+        require!(
+            transaction_account.instructions.len() == 1,
+            ErrorCode::MultipleInstructions
+        );
+
+        let instruction = &transaction_account.instructions[0];
+
+        // the instruction should be against the system program
+        transaction_account
+            .check_account(instruction.program_id, &System::id())
+            .map_err(|_| ErrorCode::InvalidProgram)?;
+
+        // Check the transaction is a transfer
+        require_eq!(
+            instruction.data[0],
+            ExecuteMiddleware::TRANSFER_INSTRUCTION_INDEX,
+            ErrorCode::InvalidInstructionType
+        );
+
+        // check that the recipient is the correct recipient
+        transaction_account
+            .check_account(
+                instruction.accounts.get(1).unwrap().key,
+                &ctx.accounts.middleware_account.recipient,
+            )
+            .map_err(|_| ErrorCode::InvalidRecipient)?;
+
+        ExecuteMiddleware::approve(ctx)
+    }
+}
+
+#[derive(Accounts)]
+#[instruction(
+/// The recipient that the middleware will allow transactions to be sent to
+recipient: Pubkey,
+/// The nonce for the middleware signer
+nonce: u8,
+/// The bump seed for the middleware signer
+bump: u8
+)]
+pub struct Create<'info> {
+    #[account(
+        init,
+        payer = authority,
+        space = 8 + 32 + 32 + 8 + 8,
+        seeds = [CheckRecipient::SEED_PREFIX, authority.key().as_ref(), nonce.to_le_bytes().as_ref()],
+        bump,
+    )]
+    pub middleware_account: Account<'info, CheckRecipient>,
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct ExecuteMiddleware<'info> {
```

Comment:
I will wait for the architecture meeting, but is the chaining element already in this version? No, correct?

Reply:
No, no chaining yet